Add setting up the secrets operators docs for secrets management (#3568)
weaveworks/weave-gitops: website/docs/secrets/setup-secrets-operator.mdx

Specially when storing Kubernetes Secrets in `Git` safely and securely.
To achive this, WGE has two approches to work with kubernetes secrets.

1- [External Secrets Operator](https://external-secrets.io/v0.8.1/)

2- [Mozilla Sops](https://fluxcd.io/flux/guides/mozilla-sops/)

## Set up ESO

WGE installs ESO as a [profile](https://github.com/weaveworks/profiles-catalog/tree/main/charts/external-secrets)

The profile consists of 2 components

1- [The HelmChart for External Secrets Operator](https://github.com/weaveworks/profiles-catalog/blob/main/charts/external-secrets/Chart.yaml)  
- This will install the operator helm chart and the operator CRDs. more info [here](https://external-secrets.io/v0.8.1/examples/gitops-using-fluxcd/)

2- [Kustomization Reference to secret stores CRs](https://github.com/weaveworks/profiles-catalog/blob/main/charts/external-secrets/templates/secret-stores-kustomization.yaml)

- This kustomization refers to the git repository with the secrets and can have the values from [values file](https://github.com/weaveworks/profiles-catalog/blob/main/charts/external-secrets/values.yaml) according to your secrets git repository

### Installation Steps

> - [Bootstrap the required secrets](/secrets/bootstraping-secrets.mdx) for the Profile Helm Repository and the Secrets Repository

Reviewer: example https://docs.gitops.weave.works/docs/policy/getting-started/#installation-pre-requisites

- Install the [Profile Helm repository](https://docs.gitops.weave.works/docs/gitops-templates/cli/#profiles)
- To install the operator on Management Cluster
    - Go to `Applications` from the side bar and `Add new application`
    - Select `External Secrets` and input your `namespace`
        ![ESO Profile](./imgs/eso1.jpg)
    - Then Adjust the `values` according to your secrets git repository
        ![ESO Values](./imgs/eso2.jpg)
    - After that Authenticate with Github, Create and merge the pull request
        ![ESO PR](./imgs/eso3.jpg)

>     - Once Flux reconcile the git source this should take few minutes, you will find the operator installed and the secrets lives in the cluster

Reviewer: ```suggestion
    - Once flux reconciles the git source, which takes a few minutes, you will find the operator installed and the secrets created in the cluster. You can now use the UI[link to ui page that ali is writting] to create secrets using your external secret operator.
```

- To bootstrap the operator on a leaf Cluster

>     - Go to `Clusters` from the side bar and `Create A Cluster`

Reviewer: ```suggestion
    - Go to `Clusters` from the side bar and  click on `Create A Cluster`.
```

    - Select your template (must have the enable profiles annotation)
        ![ESO Cluster](./imgs/eso5.jpg)

>     - Select `External Secrets` from and input your `namespace`

Reviewer: ```suggestion
    - Select `External Secrets` and input your `namespace`
```

>         ![ESO Profile](./imgs/eso4.jpg)

Reviewer: I dont feel that we need to repeat  the same steps again especially with all the screenshots and everything, after the selection point you can just say something like `configure your values and create the PR as mentioned in the previous section` and remove the last three steps.

    - Then Adjust the `values` according to your secrets git repository
        ![ESO Values](./imgs/eso2.jpg)
    - After that Authenticate with Github, Create and merge the pull request
        ![ESO PR](./imgs/eso3.jpg)
    - Once the cluster is created and Flux reconcile the git source this should take few minutes, you will find the operator installed and the secrets lives in the cluster

## Set up SOPS

- Installing Sops on Managmenet cluster, follow [this guide](/secrets/setup-sops.mdx)

### Bootstrapping Sops to leaf clusters

import SopsBootstrapJob from "!!raw-loader!./assets/sops-bootstrap-job.yaml";
import TemplateParams from "!!raw-loader!./assets/template-params.yaml";
import TemplateAnnotations from "!!raw-loader!./assets/template-annotations.yaml";

> - **This will need the following**

Reviewer: I would remove this bullter point and maye be have a prerequiste header. The bullter points in this doc have too many levels which is affecting the visibility and readability of everything.

>     - `ClusterBootstrapConfig` job to install sops and generate keys

Reviewer: ```suggestion
`ClusterBootstrapConfig` job to install sops and generate keys when bootstrapping leaf clusters. The following example is using GPG encryption.
```

>     - `ClusterBootstrapConfig` job to install sops and generate keys

Reviewer: No need for this to be a bullet point, it ends creating to many levels of bullet points. You can add a small header if you want to highlight different sections, it will make things much more readable.

>         We will need a job to run in the bootstrapping of leaf cluster to install sops and generate the keys,the following example is using GPG encryption

Reviewer: ```suggestion
```

>         - The Job CR, follow [this guide](/cluster-management/getting-started/#add-a-cluster-bootstrap-config) for more information on ClusterBootstrapConfig

Reviewer: ```suggestion
Create the following ClusterBootstrapConfig CR and push it to your fleet repo. Follow [this guide](/cluster-management/getting-started/#add-a-cluster-bootstrap-config) for more information on ClusterBootstrapConfig.
```

            <details><summary>Expand to view </summary>
                <CodeBlock
                title="clusters/management/capi/boostrap/sops-bootstrap-job.yaml"
                className="language-yaml"
                >
                    {SopsBootstrapJob}
                </CodeBlock>
            </details>

>         - Template Updates

Reviewer: Have this as a small header instead of a bullet point.

            In order to install sops we need some modifications to the template to allow it to create a [Kustomization](https://fluxcd.io/flux/guides/mozilla-sops/#configure-in-cluster-secrets-decryption) for reconciling the secrets on the cluster for sops during cluster creation for the secrets path and run the `ClusterBootstrapConfig` job

>             It will refer to `sops` location in the root of cluster files

Reviewer: this sentence seems out of context and a bit confusing, if you're talking about the kustomization, I dont feel that this is the right place to have it.

            - Template Updates
                - It should have annotation `templates.weave.works/sops-enabled: "true"` to enable WGE to create the kustomization in cluster creation

>                 - It should have the following values that’s needed for **kustomization**

Reviewer: ```suggestion
                - It should have the following values that are needed for the **kustomization**
```

>                 - GitOps template with sops enabled (check how to add the annotations below)

Reviewer: I am a bit confused, in the point before this you say "it should have the following values" but you jump straight to this next bullet point.

                    <details><summary>Expand to view </summary>
                        <CodeBlock
                        title="clusters/management/capi/boostrap/template.yaml"
                        className="language-yaml"
                        >
                        {TemplateParams}
                        </CodeBlock>
                    </details>

                - It should have the following annotations under `GitOpsCluster` to be used in the bootstrap job

                    <details><summary>Expand to view </summary>
                        <CodeBlock
                        title="clusters/management/capi/boostrap/template.yaml"
                        className="language-yaml"
                        >
                        {TemplateAnnotations}
                        </CodeBlock>
                    </details>

> - **Installation Steps**

Reviewer: This should be a header not a bullter point

    - Go to `Clusters` from the side bar and `Create A Cluster`
    - Select your Sops template (must have the annotation and parameters)
        ![ESO Cluster](./imgs/sops1.jpg)

>     - Fill the data, Create and Merge the PR

Reviewer: ```suggestion
    - Fill the data, create and merge the PR.
```

>     - You should expect the following after the cluster is ready

Reviewer: ```suggestion
    - You should expect the following after the cluster is ready:
```

      - A leaf cluster created with flux & sops bootstrapped

>       - A secret created on leaf cluster `sops-gpg` to decrypt secrets

Reviewer: I think it's better to take the screenshot with those values filled so the user can make the connection.

      - A secret created on leaf cluster `sops-gpg-pub` to encrypt secrets
      - A Kustomization with `decryption` defined in it to `sops` location in the cluster repo location
      - A PR is created to the cluster repo with the public key and sops creation rules (optional)

>       - Added Role for the public key to be accessed through management cluster

Reviewer: After the installation steps you can point the user to the UI pages so they can start using the create form

</details>

#### Cluster template updates

In order to bootstrap SOPS to leaf clusters, we need some modifications to the cluster template to allow creating a [Kustomization](https://fluxcd.io/flux/guides/mozilla-sops/#configure-in-cluster-secrets-decryption)
for reconciling the secrets on the cluster using SOPS and to run the `ClusterBootstrapConfig` job during cluster creation.

The template metadata should have annotation, it will be used by WGE to create the Kustomization with the cluster files.

```yaml
templates.weave.works/sops-enabled: "true"
```

The template should have the following parameters that are needed for the Kustomization

<details><summary>Expand to view </summary>
    <CodeBlock
    title="clusters/management/capi/templates/template.yaml"
    className="language-yaml"
    >
    {TemplateParams}
    </CodeBlock>
</details>

The template should have the following annotations under `GitOpsCluster` to be used in the bootstrap job

<details><summary>Expand to view </summary>
    <CodeBlock
    title="clusters/management/capi/templates/template.yaml"
    className="language-yaml"
    >
    {TemplateAnnotations}
    </CodeBlock>
</details>

### Installation Steps

> To bootstrap SOPS on a leaf cluster, create a new cluster using the SOPS template from the `Create Cluster` page.

Reviewer 2: I am sorry if I mistakenly said to remove the create cluster screenshot with the highlighted fields. Can we add it back in?

> To bootstrap SOPS on a leaf cluster, create a new cluster using the SOPS template from the `Create Cluster` page.

Reviewer 2: ```suggestion
To bootstrap SOPS on a leaf cluster, create a new cluster using the SOPS template from the `Create Cluster` page and fill in the following SOPS-related values in the form: 

- SOPS_KUSTOMIZATION_NAME
- SOPS_SECRET_REF
- SOPS_SECRET_REF_NAMESPACE
- SOPS_KEY_NAME
- SOPS_KEY_COMMENT
- SOPS_PUSH_TO_GIT
```

  tag: v1.0.0                                     # if you would like to download your secrets from specific tag
  # branch: main                                  # if you would like to download your secrets from specific branch
  path: .                                         # could be a path to the secrets dir or a kustomization.yaml file for the SecretStore in the GitRepository
  secretRef: ssh-creds                            # Name of the K8s secret with private repo auth credentials (private key, pub key, known hosts)
  # sourceRef:                                    # Could specify a name for an existing GitSource reference instead of creating a new one
  #   kind: GitRepository
  #   name: external-secrets
  #   namespace: external-secrets
```

## Set up SOPS

Weave GitOps Enterprise now supports managing secrets using SOPS, a tool that encrypts and decrypts secrets using various key management services.

### Install SOPS on management cluster

To install SOPS on management cluster. follow this [guide](https://fluxcd.io/flux/guides/mozilla-sops/) from flux documentation.

> #### Using SOPS via WGE UI

Reviewer: This whole section is not needed, it's a duplication of Ahmed's page.

The job is a container which generates SOPS secrets key pairs, Creates secrets with the private key, Creates a secret with the public key (to be used in self-serve flow) and the proper Rbac for it.
As well as an option to push the public key to the git repository via PR (to be distributed)

### Prerequisites

#### ClusterBootstrapConfig job

The following example is using GPG encryption to install SOPS and generate keys when bootstrapping leaf clusters. Create the following ClusterBootstrapConfig CR and push it to your fleet repo.
        
<details><summary>Expand to view </summary>
    <CodeBlock
    title="clusters/management/capi/boostrap/sops-bootstrap-job.yaml"
    className="language-yaml"
    >
        {SopsBootstrapJob}
    </CodeBlock>
</details>

#### Cluster template updates

In order to bootstrap SOPS to leaf clusters, we need some modifications to the cluster template to allow creating a [Kustomization](https://fluxcd.io/flux/guides/mozilla-sops/#configure-in-cluster-secrets-decryption)
for reconciling the secrets on the cluster using SOPS and to run the `ClusterBootstrapConfig` job during cluster creation.

> - The template metadata should have annotation `templates.weave.works/sops-enabled: "true"`. This will be used by WGE to create the Kustomization with the cluster files.

Reviewer: I would remove all the bullet points in this section, they dont add much to the explanation and are a bit confusing visually. You can just use the text as it is with code blocks under each sentence like you did.
